pkgs/applications/networking/instant-messengers/session-desktop/default.nix

```diff
@@ -18,39 +15,32 @@ let
   appimage = appimageTools.wrapType2 {
     inherit version pname src;
   };
-  appimage-contents = appimageTools.extractType2 {
+  appimageContents = appimageTools.extractType2 {
     inherit version pname src;
   };
 in
 stdenvNoCC.mkDerivation {
   inherit version pname;
   src = appimage;
 
-  nativeBuildInputs = [ copyDesktopItems makeWrapper ];
-
-  desktopItems = [
-    (makeDesktopItem {
-      name = "Session";
-      desktopName = "Session";
-      comment = "Onion routing based messenger";
-      exec = "${appimage}/bin/session-desktop-${version}";
-      icon = "${appimage-contents}/session-desktop.png";
-      terminal = false;
-      type = "Application";
-      categories = [ "Network" ];
-    })
-  ];
+  nativeBuildInputs = [ makeWrapper ];
 
   installPhase = ''
     runHook preInstall
 
-    mv bin/session-desktop-${version} bin/session-desktop
+    mv bin/${pname}-${version} bin/${pname}
 
     mkdir -p $out/
     cp -r bin $out/bin
 
-    wrapProgram $out/bin/session-desktop \
-      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}"
+    mkdir -p $out/share/${pname}
+    cp -a ${appimageContents}/usr/share/icons $out/share/
+    install -Dm 644 ${appimageContents}/${pname}.desktop -t $out/share/applications/
+
+    substituteInPlace $out/share/applications/${pname}.desktop --replace "AppRun" "${pname}"
```

Comment:
I think I woud like to keep the old desktop item as it feels more declarative and uses the full path. Compared to the upstream item that relies on the environment.

However I think it is a good idea to use the icons directly from the upstream and it would be nice if you could make that work with the current desktop item :)

Reply:
Part of the problem was that the desktop file referenced the application directly without the Wayland flags. I'll look into making the desktop file work with the wrapper, but since the flags rely on Bash variable expansion to function, I can't just embed the flag code in the desktop file.

Additionally, I do not think it's possible for the package to reference itself (correct me if I'm wrong) for the desktop file to refer to the wrapper script.

Finally, I think it's fine for the desktop file to only work if the package is installed into the path. To keep the scripts simple, I *think* it's fine to assume that the .desktop file will only be used if the package is in the path.